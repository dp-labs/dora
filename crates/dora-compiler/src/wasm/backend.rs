use dora_ir::IRTypes;
use dora_runtime::symbols;
use melior::dialect::{
    arith::{self, CmpfPredicate, CmpiPredicate},
    cf, func,
};
use melior::ir::{
    attribute::FlatSymbolRefAttribute, r#type::IntegerType, BlockRef, OperationRef, Region, Type,
    TypeLike, Value, ValueLike,
};
use wasmer_types::TrapCode;

use crate::backend::IntCC;
use crate::errors::Result;
use crate::value::ToContextValue;
use crate::{backend::TypeMethods, context::Context, conversion::builder::OpBuilder, state::State};

use super::intrinsics::WASMIntrinsics;

/// A builder structure for constructing WebAssembly (WASM) operations and managing
/// the underlying compilation process using an `OpBuilder`. The `WASMBuilder`
/// operates on a `WASMBackend` and provides an interface to emit operations
/// within the context of a given WASM execution model.
///
/// # Fields:
/// - `backend`: A mutable reference to the `WASMBackend`, which contains the overall
///   context and state of the WASM environment being built.
/// - `builder`: An `OpBuilder` instance responsible for generating operations
///   during the construction process.
///
/// # Example Usage:
/// ```no_check
/// let mut wasm_builder = WASMBuilder {
///     backend: &mut wasm_backend,
///     builder: op_builder,
/// };
/// wasm_builder.builder.some_build_method();
/// ```
///
/// # Notes:
/// - The `WASMBuilder` works closely with the `WASMBackend` to manage WebAssembly-specific
///   context and state, allowing for the construction of WASM operations within the MLIR system.
#[derive(Debug)]
pub struct WASMBuilder<'c, 'a> {
    /// A mutable reference to the WASM backend containing the context and state.
    pub backend: &'a mut WASMBackend<'c>,
    /// The builder used to generate operations within the WASM execution context.
    pub builder: OpBuilder<'c, 'a>,
}

impl<'c> std::ops::Deref for WASMBuilder<'c, '_> {
    type Target = WASMBackend<'c>;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.backend
    }
}

impl std::ops::DerefMut for WASMBuilder<'_, '_> {
    #[inline]
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.backend
    }
}

/// The backend structure for managing the WebAssembly (WASM) execution context and
/// intrinsics. `WASMBackend` encapsulates the relevant MLIR context, intrinsic types,
/// and execution state required for compiling and executing WASM bytecode.
///
/// # Fields:
/// - `ctx`: A reference to the MLIR `Context`, which provides the underlying infrastructure
///   for managing the compilation and execution process.
/// - `intrinsics`: A set of WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64,
///   that are commonly used in WASM operations.
/// - `state`: The current state of the execution, which includes the stack and control
///   structures required for managing the flow of WASM bytecode execution.
///
/// # Example Usage:
/// ```no_check
/// let wasm_backend = WASMBackend {
///     ctx: &mlir_context,
///     intrinsics: wasm_intrinsics,
///     state: wasm_state,
/// };
/// ```
///
/// # Notes:
/// - The `WASMBackend` is responsible for maintaining the state and context necessary
///   for WASM execution and compilation. It interacts with other structures like the
///   `WASMBuilder` to facilitate the construction of operations that adhere to the
///   WebAssembly execution model.
#[derive(Debug)]
pub struct WASMBackend<'c> {
    /// A reference to the MLIR context used in the WASM backend.
    pub ctx: &'c Context,
    /// WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64.
    pub intrinsics: WASMIntrinsics<'c>,
    /// The current state of WASM execution, including the stack and control structures.
    pub state: State<'c, 'c>,
}

impl<'c> WASMBackend<'c> {
    pub fn new(ctx: &'c Context) -> Self {
        Self {
            ctx,
            intrinsics: WASMIntrinsics::declare(ctx),
            state: State::new(),
        }
    }
}

impl<'c> IRTypes for WASMBackend<'c> {
    type Type = Type<'c>;
    type Value = Value<'c, 'c>;
    type Region = Region<'c>;
    type BasicBlock = BlockRef<'c, 'c>;
    type Operation = OperationRef<'c, 'c>;
}

impl TypeMethods for WASMBackend<'_> {
    fn type_ptr(&self) -> Self::Type {
        self.intrinsics.ptr_ty
    }

    fn type_int(&self, bits: u32) -> Self::Type {
        IntegerType::new(&self.ctx.mlir_context, bits).into()
    }
}

pub(crate) fn is_zero<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    value: Value<'c, 'a>,
) -> Result<Value<'c, 'a>> {
    let ty = value.r#type();
    if ty.is_integer() {
        Ok(builder
            .create(arith::cmpi(
                builder.context(),
                CmpiPredicate::Eq,
                value,
                if builder.int_ty_width(ty)? == 32 {
                    builder.create(builder.iconst_32(0)).result(0)?.into()
                } else {
                    builder.create(builder.iconst_64(0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else if ty.is_float() {
        Ok(builder
            .create(arith::cmpf(
                builder.context(),
                CmpfPredicate::Oeq,
                value,
                if ty.is_f32() {
                    builder.create(builder.fconst_32(0.0)).result(0)?.into()
                } else {
                    builder.create(builder.fconst_64(0.0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else {
        // WASM pointer type
        let value = builder.make(builder.load(value, builder.i32_ty()))?;
        let result = builder.make(builder.icmp_imm(IntCC::Equal, value, 0)?)?;
        Ok(unsafe { Value::from_raw(result.to_raw()) })
    }
}

pub(crate) fn trap<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    code: TrapCode,
    continue_block: BlockRef<'c, 'a>,
) -> Result<()> {
    let ctx = builder.ctx;
    let code = builder.make(builder.iconst_32(code as _))?;
    builder.create(func::call(
        ctx,
        FlatSymbolRefAttribute::new(ctx, symbols::wasm::RAISE_TRAP),
        &[code],
        &[],
        builder.get_insert_location(),
    ));
    builder.create(cf::br(&continue_block, &[], builder.get_insert_location()));
    Ok(())
}
