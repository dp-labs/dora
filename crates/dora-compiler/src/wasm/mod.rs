//! Reference: https://github.com/wasmerio/wasmer/tree/main/lib/compiler-llvm
#![allow(dead_code)]

mod backend;
mod code;
mod errors;
mod func;
mod intrinsics;
mod state;
mod ty;

#[cfg(test)]
mod tests;

use dora_primitives::config::OptimizationLevel;
use func::FuncTranslator;
use melior::ir::operation::OperationBuilder;
use melior::ir::{Block, Region};
use melior::ir::{Location, Module as MLIRModule};
use std::sync::Arc;
use wasmer_compiler::{
    FunctionBodyData, ModuleEnvironment, ModuleMiddleware, ModuleTranslationState,
};
use wasmer_types::entity::{EntityRef, PrimaryMap};
use wasmer_types::{
    CompileModuleInfo, Features, FunctionIndex, LocalFunctionIndex, SectionIndex, SignatureIndex,
    Symbol, SymbolRegistry,
};

use crate::errors::CompileError;
use crate::{Context, Module};

/// Represents a WebAssembly (Wasm) to LLVM/MLIR compiler. The `WASMCompiler` struct is responsible for
/// compiling WebAssembly code into LLVM/MLIR, leveraging the configuration provided at the time of
/// instantiation.
///
/// # Fields:
/// - `config`: The configuration settings for the compiler, encapsulated in the `Config` struct.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: false,
///     opt_level: OptimizationLevel::O2,
///     middlewares: vec![],
/// };
/// let wasm_compiler = WASMCompiler::new(config);
///
/// // Access the compiler's configuration
/// let compiler_config = wasm_compiler.config();
/// ```
///
/// # Notes:
/// - The `WASMCompiler` struct encapsulates all compiler functionality for transforming WebAssembly bytecode
///   into a more optimized and executable format using LLVM/MLIR.
/// - The compiler that compiles a WebAssembly module with LLVM/MLIR, translating the Wasm to MLIR IR, optimizing
///   it and then translating to assembly or executing it with JIT mode.
#[derive(Debug, Clone)]
pub struct WASMCompiler<'c> {
    /// The MLIR context used for generating operations and managing their lifetime. It encapsulates the state
    /// of the MLIR infrastructure, including types, modules, and operations. This context is tied to the
    /// lifetime `'c` of the EVMCompiler.
    pub ctx: &'c Context,
    /// The configuration settings for the Wasm compiler.
    pub config: Config,
}

impl<'c> WASMCompiler<'c> {
    /// Creates a new instance of the `WASMCompiler` with the given configuration.
    ///
    /// # Arguments:
    /// - `config`: The `Config` object that defines the compilation settings, including optimization levels
    ///   and any middleware.
    ///
    /// # Returns:
    /// - A `WASMCompiler` instance initialized with the provided configuration.
    #[inline]
    pub fn new(ctx: &'c Context, config: Config) -> WASMCompiler {
        WASMCompiler { ctx, config }
    }

    /// Retrieves the configuration for this `WASMCompiler`.
    ///
    /// # Returns:
    /// - A reference to the `Config` object that was used to configure the compiler.
    ///
    /// # Example:
    /// ```no_check
    /// let compiler_config = wasm_compiler.config();
    /// ```
    #[inline]
    fn config(&self) -> &Config {
        &self.config
    }

    /// Compile the WASM bytes using LLVM/MLIR.
    pub fn compile(&self, data: &[u8]) -> Result<Module<'c>, CompileError> {
        let environ = ModuleEnvironment::new();
        let translation = environ
            .translate(data)
            .map_err(|err| CompileError::Codegen(err.to_string()))?;
        let module = translation.module;
        let compile_info = CompileModuleInfo {
            module: Arc::new(module),
            features: Features::default(),
            memory_styles: PrimaryMap::new(),
            table_styles: PrimaryMap::new(),
        };
        self.compile_module(
            &compile_info,
            translation.module_translation_state.as_ref().unwrap(),
            translation.function_body_inputs,
        )
    }

    fn compile_module(
        &self,
        compile_info: &CompileModuleInfo,
        module_translation: &ModuleTranslationState,
        function_body_inputs: PrimaryMap<LocalFunctionIndex, FunctionBodyData<'_>>,
    ) -> Result<Module<'c>, CompileError> {
        // WASM Module information
        let memory_styles = &compile_info.memory_styles;
        let table_styles = &compile_info.table_styles;
        let wasm_module = &compile_info.module;
        let function_body_inputs = function_body_inputs
            .iter()
            .collect::<Vec<(LocalFunctionIndex, &FunctionBodyData<'_>)>>();
        let functions: Vec<_> = function_body_inputs
            .iter()
            .map(|(i, input)| {
                FuncTranslator::translate(
                    self.ctx,
                    &self.config,
                    wasm_module,
                    module_translation,
                    i,
                    input,
                    memory_styles,
                    table_styles,
                    &ShortNames {},
                )
            })
            .collect();
        // MLIR Module region and block
        let module_block = Block::new(&[]);
        for func in functions {
            module_block
                .append_operation(func.map_err(|err| CompileError::Codegen(err.to_string()))?);
        }
        let module_region = Region::new();
        module_region.append_block(module_block);
        // Build main module
        let op = OperationBuilder::new("builtin.module", Location::unknown(&self.ctx.mlir_context))
            .add_regions([module_region])
            .build()?;
        Ok(Module::new(
            MLIRModule::from_operation(op).expect("module failed to create"),
        ))
    }
}

struct ShortNames {}

impl SymbolRegistry for ShortNames {
    fn symbol_to_name(&self, symbol: Symbol) -> String {
        match symbol {
            Symbol::Metadata => "M".to_string(),
            Symbol::LocalFunction(index) => format!("f{}", index.index()),
            Symbol::Section(index) => format!("s{}", index.index()),
            Symbol::FunctionCallTrampoline(index) => format!("t{}", index.index()),
            Symbol::DynamicFunctionTrampoline(index) => format!("d{}", index.index()),
        }
    }

    fn name_to_symbol(&self, name: &str) -> Option<Symbol> {
        if name.len() < 2 {
            return None;
        }
        let (ty, idx) = name.split_at(1);
        if ty.starts_with('M') {
            return Some(Symbol::Metadata);
        }

        let idx = idx.parse::<u32>().ok()?;
        match ty.chars().next().unwrap() {
            'f' => Some(Symbol::LocalFunction(LocalFunctionIndex::from_u32(idx))),
            's' => Some(Symbol::Section(SectionIndex::from_u32(idx))),
            't' => Some(Symbol::FunctionCallTrampoline(SignatureIndex::from_u32(
                idx,
            ))),
            'd' => Some(Symbol::DynamicFunctionTrampoline(FunctionIndex::from_u32(
                idx,
            ))),
            _ => None,
        }
    }
}

/// Represents the configuration options for the `WASMCompiler`. The `Config` struct defines various
/// settings that influence the compilation process, such as optimization levels and verification options.
///
/// # Fields:
/// - `enable_nan_canonicalization`: A boolean flag indicating whether to enable NaN canonicalization, which
///   standardizes NaN representations in floating-point operations.
/// - `enable_verifier`: A boolean flag to enable or disable the verification phase during compilation, ensuring
///   that the compiled code meets correctness constraints.
/// - `opt_level`: The optimization level to be applied during compilation, which can affect the performance and
///   size of the compiled output.
/// - `middlewares`: A vector of middleware components that modify or extend the behavior of the compilation
///   process. Each middleware is represented as a trait object implementing the `ModuleMiddleware` trait.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: true,
///     opt_level: OptimizationLevel::O3,
///     middlewares: vec![],
/// };
/// ```
///
/// # Notes:
/// - The `Config` struct is essential for controlling the behavior of the `WASMCompiler`. You can tune the
///   compilation process by adjusting these fields based on the requirements of your WebAssembly module.
#[derive(Debug, Clone, Default)]
pub struct Config {
    /// A flag to enable NaN canonicalization during floating-point operations.
    pub(crate) enable_nan_canonicalization: bool,
    /// A flag to enable or disable the verifier during compilation.
    pub(crate) enable_verifier: bool,
    /// The optimization level for the compilation process.
    pub(crate) opt_level: OptimizationLevel,
    /// A collection of middleware components that modify the behavior of the compiler.
    pub(crate) middlewares: Vec<Arc<dyn ModuleMiddleware>>,
}
