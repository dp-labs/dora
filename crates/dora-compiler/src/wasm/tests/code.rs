//！Reference: https://github.com/WebAssembly/spec/tree/main/test/core
use crate::context::Context;
use crate::wasm::{Config, WASMCompiler};
use wasmer::wat2wasm;

macro_rules! assert_snapshot {
    ($code:expr) => {
        let context = Context::new();
        let compiler = WASMCompiler::new(&context, Config::default());
        let wasm_bytes = wat2wasm($code.as_bytes()).unwrap();
        // Compile EVM Bytecode to MLIR EVM Dialect
        let module = compiler
            .compile(&wasm_bytes)
            .expect("failed to compile program");
        let op = module.module().as_operation();
        assert!(op.verify());
        insta::assert_snapshot!(op);
    };
}

#[test]
fn sum() {
    assert_snapshot!(
        r#"
(module
  (type $sum_t (func (param i32 i32) (result i32)))
  (func $sum_f (type $sum_t) (param $x i32) (param $y i32) (result i32)
    local.get $x
    local.get $y
    i32.add)
  (export "sum" (func $sum_f)))
"#
    );
}

#[test]
fn i32_arith() {
    assert_snapshot!(
        r#"
(module
  (func (export "add") (param $x i32) (param $y i32) (result i32) (i32.add (local.get $x) (local.get $y)))
  (func (export "sub") (param $x i32) (param $y i32) (result i32) (i32.sub (local.get $x) (local.get $y)))
  (func (export "mul") (param $x i32) (param $y i32) (result i32) (i32.mul (local.get $x) (local.get $y)))
  (func (export "div_s") (param $x i32) (param $y i32) (result i32) (i32.div_s (local.get $x) (local.get $y)))
  (func (export "div_u") (param $x i32) (param $y i32) (result i32) (i32.div_u (local.get $x) (local.get $y)))
  (func (export "rem_s") (param $x i32) (param $y i32) (result i32) (i32.rem_s (local.get $x) (local.get $y)))
  (func (export "rem_u") (param $x i32) (param $y i32) (result i32) (i32.rem_u (local.get $x) (local.get $y)))
  (func (export "and") (param $x i32) (param $y i32) (result i32) (i32.and (local.get $x) (local.get $y)))
  (func (export "or") (param $x i32) (param $y i32) (result i32) (i32.or (local.get $x) (local.get $y)))
  (func (export "xor") (param $x i32) (param $y i32) (result i32) (i32.xor (local.get $x) (local.get $y)))
  (func (export "shl") (param $x i32) (param $y i32) (result i32) (i32.shl (local.get $x) (local.get $y)))
  (func (export "shr_s") (param $x i32) (param $y i32) (result i32) (i32.shr_s (local.get $x) (local.get $y)))
  (func (export "shr_u") (param $x i32) (param $y i32) (result i32) (i32.shr_u (local.get $x) (local.get $y)))
  (func (export "rotl") (param $x i32) (param $y i32) (result i32) (i32.rotl (local.get $x) (local.get $y)))
  (func (export "rotr") (param $x i32) (param $y i32) (result i32) (i32.rotr (local.get $x) (local.get $y)))
  (func (export "clz") (param $x i32) (result i32) (i32.clz (local.get $x)))
  (func (export "ctz") (param $x i32) (result i32) (i32.ctz (local.get $x)))
  (func (export "popcnt") (param $x i32) (result i32) (i32.popcnt (local.get $x)))
  (func (export "extend8_s") (param $x i32) (result i32) (i32.extend8_s (local.get $x)))
  (func (export "extend16_s") (param $x i32) (result i32) (i32.extend16_s (local.get $x)))
  (func (export "eqz") (param $x i32) (result i32) (i32.eqz (local.get $x)))
  (func (export "eq") (param $x i32) (param $y i32) (result i32) (i32.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x i32) (param $y i32) (result i32) (i32.ne (local.get $x) (local.get $y)))
  (func (export "lt_s") (param $x i32) (param $y i32) (result i32) (i32.lt_s (local.get $x) (local.get $y)))
  (func (export "lt_u") (param $x i32) (param $y i32) (result i32) (i32.lt_u (local.get $x) (local.get $y)))
  (func (export "le_s") (param $x i32) (param $y i32) (result i32) (i32.le_s (local.get $x) (local.get $y)))
  (func (export "le_u") (param $x i32) (param $y i32) (result i32) (i32.le_u (local.get $x) (local.get $y)))
  (func (export "gt_s") (param $x i32) (param $y i32) (result i32) (i32.gt_s (local.get $x) (local.get $y)))
  (func (export "gt_u") (param $x i32) (param $y i32) (result i32) (i32.gt_u (local.get $x) (local.get $y)))
  (func (export "ge_s") (param $x i32) (param $y i32) (result i32) (i32.ge_s (local.get $x) (local.get $y)))
  (func (export "ge_u") (param $x i32) (param $y i32) (result i32) (i32.ge_u (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn i64_arith() {
    assert_snapshot!(
        r#"
(module
  (func (export "add") (param $x i64) (param $y i64) (result i64) (i64.add (local.get $x) (local.get $y)))
  (func (export "sub") (param $x i64) (param $y i64) (result i64) (i64.sub (local.get $x) (local.get $y)))
  (func (export "mul") (param $x i64) (param $y i64) (result i64) (i64.mul (local.get $x) (local.get $y)))
  (func (export "div_s") (param $x i64) (param $y i64) (result i64) (i64.div_s (local.get $x) (local.get $y)))
  (func (export "div_u") (param $x i64) (param $y i64) (result i64) (i64.div_u (local.get $x) (local.get $y)))
  (func (export "rem_s") (param $x i64) (param $y i64) (result i64) (i64.rem_s (local.get $x) (local.get $y)))
  (func (export "rem_u") (param $x i64) (param $y i64) (result i64) (i64.rem_u (local.get $x) (local.get $y)))
  (func (export "and") (param $x i64) (param $y i64) (result i64) (i64.and (local.get $x) (local.get $y)))
  (func (export "or") (param $x i64) (param $y i64) (result i64) (i64.or (local.get $x) (local.get $y)))
  (func (export "xor") (param $x i64) (param $y i64) (result i64) (i64.xor (local.get $x) (local.get $y)))
  (func (export "shl") (param $x i64) (param $y i64) (result i64) (i64.shl (local.get $x) (local.get $y)))
  (func (export "shr_s") (param $x i64) (param $y i64) (result i64) (i64.shr_s (local.get $x) (local.get $y)))
  (func (export "shr_u") (param $x i64) (param $y i64) (result i64) (i64.shr_u (local.get $x) (local.get $y)))
  (func (export "rotl") (param $x i64) (param $y i64) (result i64) (i64.rotl (local.get $x) (local.get $y)))
  (func (export "rotr") (param $x i64) (param $y i64) (result i64) (i64.rotr (local.get $x) (local.get $y)))
  (func (export "clz") (param $x i64) (result i64) (i64.clz (local.get $x)))
  (func (export "ctz") (param $x i64) (result i64) (i64.ctz (local.get $x)))
  (func (export "popcnt") (param $x i64) (result i64) (i64.popcnt (local.get $x)))
  (func (export "extend8_s") (param $x i64) (result i64) (i64.extend8_s (local.get $x)))
  (func (export "extend16_s") (param $x i64) (result i64) (i64.extend16_s (local.get $x)))
  (func (export "extend32_s") (param $x i64) (result i64) (i64.extend32_s (local.get $x)))
  (func (export "eqz") (param $x i64) (result i32) (i64.eqz (local.get $x)))
  (func (export "eq") (param $x i64) (param $y i64) (result i32) (i64.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x i64) (param $y i64) (result i32) (i64.ne (local.get $x) (local.get $y)))
  (func (export "lt_s") (param $x i64) (param $y i64) (result i32) (i64.lt_s (local.get $x) (local.get $y)))
  (func (export "lt_u") (param $x i64) (param $y i64) (result i32) (i64.lt_u (local.get $x) (local.get $y)))
  (func (export "le_s") (param $x i64) (param $y i64) (result i32) (i64.le_s (local.get $x) (local.get $y)))
  (func (export "le_u") (param $x i64) (param $y i64) (result i32) (i64.le_u (local.get $x) (local.get $y)))
  (func (export "gt_s") (param $x i64) (param $y i64) (result i32) (i64.gt_s (local.get $x) (local.get $y)))
  (func (export "gt_u") (param $x i64) (param $y i64) (result i32) (i64.gt_u (local.get $x) (local.get $y)))
  (func (export "ge_s") (param $x i64) (param $y i64) (result i32) (i64.ge_s (local.get $x) (local.get $y)))
  (func (export "ge_u") (param $x i64) (param $y i64) (result i32) (i64.ge_u (local.get $x) (local.get $y)))
)"#
    );
}

#[test]
fn f32_arith() {
    assert_snapshot!(
        r#"
(module
  (func (export "add") (param $x f32) (param $y f32) (result f32) (f32.add (local.get $x) (local.get $y)))
  (func (export "sub") (param $x f32) (param $y f32) (result f32) (f32.sub (local.get $x) (local.get $y)))
  (func (export "mul") (param $x f32) (param $y f32) (result f32) (f32.mul (local.get $x) (local.get $y)))
  (func (export "div") (param $x f32) (param $y f32) (result f32) (f32.div (local.get $x) (local.get $y)))
  (func (export "sqrt") (param $x f32) (result f32) (f32.sqrt (local.get $x)))
  (func (export "min") (param $x f32) (param $y f32) (result f32) (f32.min (local.get $x) (local.get $y)))
  (func (export "max") (param $x f32) (param $y f32) (result f32) (f32.max (local.get $x) (local.get $y)))
  (func (export "ceil") (param $x f32) (result f32) (f32.ceil (local.get $x)))
  (func (export "floor") (param $x f32) (result f32) (f32.floor (local.get $x)))
  (func (export "trunc") (param $x f32) (result f32) (f32.trunc (local.get $x)))
  (func (export "nearest") (param $x f32) (result f32) (f32.nearest (local.get $x)))

  (func (export "abs") (param $x f32) (result f32) (f32.abs (local.get $x)))
  (func (export "neg") (param $x f32) (result f32) (f32.neg (local.get $x)))
  (func (export "copysign") (param $x f32) (param $y f32) (result f32) (f32.copysign (local.get $x) (local.get $y)))

  (func (export "eq") (param $x f32) (param $y f32) (result i32) (f32.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x f32) (param $y f32) (result i32) (f32.ne (local.get $x) (local.get $y)))
  (func (export "lt") (param $x f32) (param $y f32) (result i32) (f32.lt (local.get $x) (local.get $y)))
  (func (export "le") (param $x f32) (param $y f32) (result i32) (f32.le (local.get $x) (local.get $y)))
  (func (export "gt") (param $x f32) (param $y f32) (result i32) (f32.gt (local.get $x) (local.get $y)))
  (func (export "ge") (param $x f32) (param $y f32) (result i32) (f32.ge (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn f64_arith() {
    assert_snapshot!(
        r#"
(module
  (func (export "add") (param $x f64) (param $y f64) (result f64) (f64.add (local.get $x) (local.get $y)))
  (func (export "sub") (param $x f64) (param $y f64) (result f64) (f64.sub (local.get $x) (local.get $y)))
  (func (export "mul") (param $x f64) (param $y f64) (result f64) (f64.mul (local.get $x) (local.get $y)))
  (func (export "div") (param $x f64) (param $y f64) (result f64) (f64.div (local.get $x) (local.get $y)))
  (func (export "sqrt") (param $x f64) (result f64) (f64.sqrt (local.get $x)))
  (func (export "min") (param $x f64) (param $y f64) (result f64) (f64.min (local.get $x) (local.get $y)))
  (func (export "max") (param $x f64) (param $y f64) (result f64) (f64.max (local.get $x) (local.get $y)))
  (func (export "ceil") (param $x f64) (result f64) (f64.ceil (local.get $x)))
  (func (export "floor") (param $x f64) (result f64) (f64.floor (local.get $x)))
  (func (export "trunc") (param $x f64) (result f64) (f64.trunc (local.get $x)))
  (func (export "nearest") (param $x f64) (result f64) (f64.nearest (local.get $x)))

  (func (export "abs") (param $x f64) (result f64) (f64.abs (local.get $x)))
  (func (export "neg") (param $x f64) (result f64) (f64.neg (local.get $x)))
  (func (export "copysign") (param $x f64) (param $y f64) (result f64) (f64.copysign (local.get $x) (local.get $y)))

  (func (export "eq") (param $x f64) (param $y f64) (result i32) (f64.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x f64) (param $y f64) (result i32) (f64.ne (local.get $x) (local.get $y)))
  (func (export "lt") (param $x f64) (param $y f64) (result i32) (f64.lt (local.get $x) (local.get $y)))
  (func (export "le") (param $x f64) (param $y f64) (result i32) (f64.le (local.get $x) (local.get $y)))
  (func (export "gt") (param $x f64) (param $y f64) (result i32) (f64.gt (local.get $x) (local.get $y)))
  (func (export "ge") (param $x f64) (param $y f64) (result i32) (f64.ge (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn global_get_and_set() {
    assert_snapshot!(
        r#"
(module
  (global $a i32 (i32.const -2))
  (global $b i64 (i64.const -5))
  (global $x (mut i32) (i32.const -12))
  (global $y (mut i64) (i64.const -15))

  (func (export "get-a") (result i32) (global.get $a))
  (func (export "get-b") (result i64) (global.get $b))
  (func (export "set-x") (param i32) (global.set $x (local.get 0)))
  (func (export "set-y") (param i64) (global.set $y (local.get 0)))
)
"#
    );
}

#[test]
fn local_get_and_set() {
    assert_snapshot!(
        r#"
(module
  (func (export "type-local-i32") (result i32) (local i32) 
    (local.set 0 (i32.const 42))
    (local.get 0)
  )
  (func (export "type-local-i64") (result i64) (local i64) 
    (local.set 0 (i64.const 42))
    (local.get 0)
  )
  (func (export "type-local-f32") (result f32) (local f32) 
    (local.set 0 (f32.const 42.0))
    (local.get 0)
  )
  (func (export "type-local-f64") (result f64) (local f64) 
    (local.set 0 (f64.const 42.0))
    (local.get 0)
  )
)
"#
    );
}

#[test]
fn local_call() {
    assert_snapshot!(
        r#"
(module
    (func $add (param $a i32) (param $b i32) (result i32)
        local.get $a
        local.get $b
        i32.add
    )
    (func $main (result i32)
        i32.const 2
        i32.const 3
        call $add
    )
)
"#
    );
}

#[test]
fn select() {
    assert_snapshot!(
        r#"
(module
    (func $select (param $cond i32) (param $a i32) (param $b i32) (result i32)
        local.get $cond
        local.get $a
        local.get $b
        select
    )
    (export "select" (func $select))
)
"#
    );
}
