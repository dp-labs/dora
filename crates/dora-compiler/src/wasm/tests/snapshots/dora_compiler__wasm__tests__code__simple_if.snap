---
source: crates/dora-compiler/src/wasm/tests/code.rs
expression: op
snapshot_kind: text
---
module {
  func.func @func1(%arg0: !llvm.ptr) {
    %0 = llvm.getelementptr %arg0[32] : (!llvm.ptr) -> !llvm.ptr, i8
    %1 = llvm.load %0 : !llvm.ptr -> !llvm.ptr
    %2 = llvm.getelementptr %arg0[8] : (!llvm.ptr) -> !llvm.ptr, i8
    %3 = llvm.getelementptr %2[0] : (!llvm.ptr) -> !llvm.ptr, !llvm.ptr
    %4 = llvm.load %3 : !llvm.ptr -> !llvm.ptr
    cf.br ^bb1
  ^bb1:  // pred: ^bb0
    %c0_i32 = arith.constant 0 : i32
    %c0_i32_0 = arith.constant 0 : i32
    %5 = "wasm.eq"(%c0_i32, %c0_i32_0) : (i32, i32) -> i32
    %c0_i32_1 = arith.constant 0 : i32
    %6 = arith.cmpi ne, %5, %c0_i32_1 : i32
    cf.cond_br %6, ^bb3, ^bb4
  ^bb2:  // 2 preds: ^bb3, ^bb5
    return
  ^bb3:  // pred: ^bb1
    %7 = llvm.load %1 : !llvm.ptr -> i32
    %c0_i32_2 = arith.constant 0 : i32
    %c0_i64 = arith.constant 0 : i64
    %8 = arith.extui %7 : i32 to i64
    %9 = arith.addi %8, %c0_i64 : i64
    %10 = llvm.getelementptr %4[%9] : (!llvm.ptr, i64) -> !llvm.ptr, !llvm.ptr
    %11 = arith.trunci %c0_i32_2 : i32 to i8
    llvm.store %11, %10 : i8, !llvm.ptr
    cf.br ^bb2
  ^bb4:  // pred: ^bb1
    cf.br ^bb5
  ^bb5:  // pred: ^bb4
    %12 = llvm.load %1 : !llvm.ptr -> i32
    cf.br ^bb2
  }
  func.func private @dora_fn_wasm_table_init(!llvm.ptr, i32, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_table_fill(!llvm.ptr, i32, i32, !llvm.ptr, i32)
  func.func private @dora_fn_wasm_table_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_table_get(!llvm.ptr, i32, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_table_set(!llvm.ptr, i32, i32, !llvm.ptr)
  func.func private @dora_fn_wasm_imported_table_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_imported_table_get(!llvm.ptr, i32, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_imported_table_grow(!llvm.ptr, !llvm.ptr, i32, i32) -> i32
  func.func private @dora_fn_wasm_memory_init(!llvm.ptr, i32, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_memory_fill(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_memory_notify(!llvm.ptr, i32, i32, i32) -> i32
  func.func private @dora_fn_wasm_memory_wait32(!llvm.ptr, i32, i32, i32, i64) -> i32
  func.func private @dora_fn_wasm_memory_wait64(!llvm.ptr, i32, i32, i64, i64) -> i32
  func.func private @dora_fn_wasm_imported_memory_copy(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_imported_memory_fill(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_imported_memory_notify(!llvm.ptr, i32, i32, i32) -> i32
  func.func private @dora_fn_wasm_imported_memory_wait32(!llvm.ptr, i32, i32, i32, i64) -> i32
  func.func private @dora_fn_wasm_imported_memory_wait64(!llvm.ptr, i32, i32, i64, i64) -> i32
  func.func private @dora_fn_wasm_func_ref(!llvm.ptr, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_data_drop(!llvm.ptr, i32)
  func.func private @dora_fn_wasm_elem_drop(!llvm.ptr, i32)
  func.func private @dora_fn_wasm_raise_trap(i32)
}
